ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/consumer/CatchupConsumer.java

```diff
@@ -0,0 +1,248 @@
+      final LogicalSchema logicalSchema,
+      final KafkaConsumer<Object, GenericRow> consumer,
+      final Supplier<LatestConsumer> latestConsumerSupplier,
+      final CatchupCoordinator catchupCoordinator,
+      final PushOffsetRange offsetRange,
+      final Clock clock,
+      final Consumer<Long> sleepMs
+  ) {
+    super(topicName, windowed, logicalSchema, consumer, clock);
+    this.latestConsumerSupplier = latestConsumerSupplier;
+    this.catchupCoordinator = catchupCoordinator;
+    this.offsetRange = offsetRange;
+    this.sleepMs = sleepMs;
+  }
+
+  public CatchupConsumer(
+      final String topicName,
+      final boolean windowed,
+      final LogicalSchema logicalSchema,
+      final KafkaConsumer<Object, GenericRow> consumer,
+      final Supplier<LatestConsumer> latestConsumerSupplier,
+      final CatchupCoordinator catchupCoordinator,
+      final PushOffsetRange offsetRange,
+      final Clock clock
+  ) {
+    this(topicName, windowed, logicalSchema, consumer, latestConsumerSupplier, catchupCoordinator,
+        offsetRange, clock, CatchupConsumer::sleep);
+  }
+
+  public interface CatchupConsumerFactory {
+    CatchupConsumer create(
+        String topicName,
+        boolean windowed,
+        LogicalSchema logicalSchema,
+        KafkaConsumer<Object, GenericRow> consumer,
+        Supplier<LatestConsumer> latestConsumerSupplier,
+        CatchupCoordinator catchupCoordinator,
+        PushOffsetRange pushOffsetRange,
+        Clock clock
+    );
+  }
+
+  @Override
+  protected void onEmptyRecords() {
+    checkCaughtUp();
+  }
+
+  @Override
+  protected void afterCommit() {
+    checkCaughtUp();
+  }
+
+  /**
+   * Called when a new assignment has been made with a call to {@link #newAssignment(Collection)}.
+   * If it's being called manually, a mapping of starting offsets is given.  This ensures that we
+   * take our current position to be the one we seek to rather than the last committed offsets.
+   * @param startingOffsets The starting offsets to use, if any.  Otherwise, we use the last
+   *     committed.
+   */
+  protected void onNewAssignment(final Optional<Map<Integer, Long>> startingOffsets) {
+    final Set<TopicPartition> tps = waitForNewAssignmentFromLatestConsumer();
+
+    consumer.assign(tps);
+    updateCurrentPositions(startingOffsets);
+    newAssignment = false;
+  }
+
+  /**
+   * Called when a new assignment has been made with a call to {@link #newAssignment(Collection)}.
+   * Assumes it should take the current position to be the last committed offsets.
+   */
+  @Override
+  protected void onNewAssignment() {
+    onNewAssignment(Optional.empty());
+  }
+
+  /**
+   * The first time we start up a LatestConsumer, if it doesn't already have an assignment, we need
+   * to wait for it to give us that assignment by calling {@link #newAssignment}
+   * @return The assignment given by latest
+   */
+  protected Set<TopicPartition> waitForNewAssignmentFromLatestConsumer() {
```

Comment:
Obviously, I'm not deeply familiar with this code (yet), but the relationship between the catchup and latest consumers looks strange to me. It seems like the latest consumer could "simply" manage the state of the catchup during rebalances and we would then not need any synchronization, atomic references, polling for assignments or state changes, etc. Maybe I'm missing something, though.

Coming back later with an edit: I'm also wondering if having separate threads for the latest and catchup consumers is worth all the hassle. It seems like we could ditch a lot of the complexity here by just taking turns polling them in the same thread, which would also be the same thread that manages all the assignment transitions. E.g., I don't think we'd need the stutter mechanism at all. Again, I might just be missing something here.

Reply:
Sounds like a good plan to me. This isn't the kind of idea worth stopping the train for, but good to keep in mind if things get hard to manage in practice.